### ISSUE TITLE
[6.x] Engagements using default background, follow default background option changes.

### ISSUE THREAD

**Issue by eri-trabiccolo** [gocodebox/lifterlms#2032]
### Reproduction steps
- create a certificate template using default background.
- award that certificate.
- now change the default background settings.

### Expected behavior
- I would expect both the published template (not totally sure on this one) AND (mostly) the awarded certificate retain the "old" default background image.

Our rule has always been that an awarded certificate, once "issued" cannot dynamically change.
As said I'm not totally sure that a template's default background should not follow the global setting change, although I'm inclined to think so, but, IMHO, for sure an award should not:
- if the award has a parent template one should change the template and then sync its awards.
- if the award has no parent template then the change should be manually done.

The implication of this is that when an engagement is created with the default background, the default background becomes its featured image (which yeah, means that the default backgrounds must be media).

What do you think @gocodebox/engineering ? 

### Actual behavior
- Both the published template and the awarded certificate to it related show the "new" default background image.

**Comment by thomasplevy** [comment]
@eri-trabiccolo this is "working as designed" though I'm open to changing the behavior. However, I'd like to note that *this is the way it works on 5.x currently. When creating (or awarding) a certificate, the background image is *not* attached as a file from the media library. We'd need to programmatically insert a new media lib object every time the default changes and then automatically add that as the featured image every time a certificate (template or award) is created.

None of this is *impossible* but this is the way it works currently and to my knowledge, outside of this bug report, we've never had a user complaint about this behavior.

**Comment by pondermatic** [comment]
I feel that some users may expect that changing the default certificate background will also change it on awarded certificates that use the default. There are probably some users that expect awarded certificates that use the default background to use the same image that was the default when it was awarded.
![emoticon-smileys-split](https://user-images.githubusercontent.com/5377968/156000326-bc3cfdd0-6c24-40c8-ad52-3421eca5b4e6.jpg)

**Comment by eri-trabiccolo** [comment]
@thomasplevy 
> @eri-trabiccolo this is "working as designed" though I'm open to changing the behavior. However, I'd like to note that *this is the way it works on 5.x currently. When creating (or awarding) a certificate, the background image is _not_ attached as a file from the media library. We'd need to programmatically insert a new media lib object every time the default changes and then automatically add that as the featured image every time a certificate (template or award) is created.

Yeah this is working as designed, and yeah this is they way it "technically" works on 5.x currently. The thing is that there was no option to change the default background before so nobody would ever notice that.

> We'd need to programmatically insert a new media lib object every time the default changes.

Not really. We just need to insert OUR defaults change, meaning that we need to create media for the old backgrounds, and the new ones we're going to ship in 6.0.
When the users change the default backgrounds from the engagements settings they already:
- use an already uploaded media
- upload an image and the then use it 


> None of this is _impossible_ but this is the way it works currently and to my knowledge, outside of this bug report, we've never had a user complaint about this behavior.

As said above, nobody ever changed the default though (if not with the filter), there was no option to do that, so it's pretty normal that nobody complaint about that.

@pondermatic

> I feel that some users may expect that changing the default certificate background will also change it on awarded certificates that use the default. There are probably some users that expect awarded certificates that use the default background to use the same image that was the default when it was awarded. 

I'm not sure about this. As I said I'm not totally sure if it's expected that a template, which uses the default background, inherits it from settings even after it has been created, but I'd say that I, as user, would not expect that the awarded certificates do the same, that's because I might have liked a default background when I set up the template and decided that was fine, but then if the software developer changes it and I don't like it I should be forced to go change all the awards?

**Comment by pondermatic** [comment]
@eri-trabiccolo wrote,
> I should be forced to go change all the awards?

How about bulk sync?

**Comment by eri-trabiccolo** [comment]
@pondermatic 
Yeah, I didn't mean necessarily one by one. I mean, why I, user,  should be forced to change them because of the plugin deciding to change the defaults? 
I think that we took for granted that defaults means "always inherit from global settings", but that's not implicit. Defaults can be "create inheriting from global settings" and that's it.
Additionally, awards without parents cannot be synced, so yeah, they must be changed 1by1.
Also I could reply with the opposite:
Please you plugin dev don't alter what I already created, then I will, if I want, change the cert template I want and sync its awards.

**Comment by pondermatic** [comment]
Sounds like if we were to add the ability to update awarded certs when the default cert background changes, we should make it optional. Maybe on the "Certificate Settings" area of the LifterLMS Settings -> Engagements tab?

**Comment by eri-trabiccolo** [comment]
Might make sense to have a "Sync" button there that spawns a bg process.
But really, my opinion is that nobody expects such a thing to be happen automatically.
Let me rephrase it: I wouldn't expect that.
I think that customers who opted for using the default background did it b/c of it somewhat fitting their style, their brand. Some of them might have not changed it just for laziness, and I doubt they would complain for not updating them with the new defaults.

**Comment by thomasplevy** [comment]
@gocodebox/engineering 

What about addressing this by describing exactly how it works on the admin panel underneath the default option selector:

"Used as the certificate background image for any certificate templates or awards which do not specify their own background image. Changing this background image will automatically update any existing certificate template or award which does not have a background image specified._"

Again... I'm not opposed to changing the existing behavior but given that we can't internally decide on what the *expected behavior should be* I'm leaning towards leaving the behavior as it was on 5.x and lower and just making it clear to users what they're getting themselves involved in by changing.

Thoughts?

**Comment by pondermatic** [comment]
@thomasplevy,
I like your idea of explaining the current behavior to the user. If anyone requests a different behavior, we'll address it then.

**Comment by eri-trabiccolo** [comment]
> What about addressing this by describing exactly how it works on the admin panel underneath the default option selector:
> 
> "Used as the certificate background image for any certificate templates or awards which do not specify their own background image. Changing this background image will automatically update any existing certificate template or award which does not have a background image specified._"

I could be fine with this for new changes to the defaults made by the users.
But when the users will update to 6.0, as of now, all of their engagements with `default` bg will show the new ones... Am I wrong? I'm not sure this is _expected_

> Again... I'm not opposed to changing the existing behavior but given that we can't internally decide on what the expected behavior should be I'm leaning towards leaving the behavior as it was on 5.x and lower and just making it clear to users what they're getting themselves involved in by changing.

I'm sorry but when you say `existing behavior` you, me and James, read the code and know you are right, but from the regular customer standpoint the default behavior is just something they never bumped into, as the only way to change the global `default` backgrounds was by using a filter: customers who used that filter were then informed of what would have happened, and they did it on purpose, I mean they had to do something to make it happen.

**Comment by thomasplevy** [comment]
> I could be fine with this for new changes to the defaults made by the users.
But when the users will update to 6.0, as of now, all of their engagements with default bg will show the new ones... Am I wrong? I'm not sure this is expected

This is addressed in #2030 

Existing behavior

Just because they haven't bumped into it doesn't mean it *changes* the existing behavior is any different.

If we fix #2030 and then *TELL* the users how the behavior works you aren't satisfied? Can you give me your concise suggestion on how to address this then please.

**Comment by eri-trabiccolo** [comment]
> Just because they haven't bumped into it doesn't mean it changes the existing behavior is any different.

I was saying that saying "that's the existing behavior" doesn't add anything to this conversation since I'm talking about the user experience and the existing behavior wasn't something the user experienced prior to 6.0.
But I understand that probably the discussion on the existing behavior generated from my:
> Our rule has always been that an awarded certificate, once "issued" cannot dynamically change.

Yeah that's what I think it was their perception.

> If we fix https://github.com/gocodebox/lifterlms/issues/2030 and then TELL the users how the behavior works you aren't satisfied?

I will be satisfied. 
I forgot about the #2030 being a separate issue and was treating this and that as a whole.

> Can you give me your concise suggestion on how to address this then please.

Sure, even if I'm satisfied now:
- on updating to 6.0 create 4 new media: the old backgrounds and the new ones.
- every time an award or an award template is created if it has a featured image then nothing to do, otherwise grab the default background's media id and set it as featured image id.

**Comment by thomasplevy** [comment]
I'm torn on what the best approach is here... so let's put it to a vote:

A) Store the default image as the featured image when creating awards and award templates

> every time an award or an award template is created if it has a featured image then nothing to do, otherwise grab the default background's media id and set it as featured image id.

In this scenario it will not be possible to blanket change the background image of awards. If you change the image you'll have to manually update the image across all awards / templates that use the image. Templates -> awards is easy because we already have bulk syncing in place.

We could additionally add a sync from the options panel. However, doing so means we also need to store additional metadata on the awards that states it is using the default image so that when the image changes and a sync is run we can identify what to change.

B) Don't store default images as the featured image and always pull the image from the option. Update language to reflect behavior: 

> "Used as the certificate background image for any certificate templates or awards which do not specify their own background image. Changing this background image will automatically update any existing certificate template or award which does not have a background image specified._"

In this scenario it is easy to bulk update awards / templates using the default image. However, it becomes difficult to update this while retaining the default on existing awards / templates. For example, I use the default for 20 templates and then want to change it for new templates. I need to first update the 20 existing templates (and then sync to the awards) and then change the default.

**Comment by eri-trabiccolo** [comment]
I'm not sure either.
I think that my opinion is always that something created with certain defaults should retain them, why shouldn't they?
If I change the default size of a certificate are templates/awards affected by this change? No. Ok you can update the language but it just makes no sense to me.

I'm for A)
and yes this probably means we have to store in a meta when engagements are using default bg or not, for future use, like if we want to build that "sync" tool.

**Comment by thomasplevy** [comment]
I really don't like option A because the amount of additional code we need to introduce seems overwhelming to me at this stage in the release cycle. I want to release. I don't want to redesign... But if this is the way we need to do it then so be it... 

Noting so I don't forget: we'll need to store metadata on post creation and then remove that metadata when the image is updated.

Also... A further complication I'm thinking about it how to handle syncing scenarios after multiple changes... by this I mean:

The sync from the options page will sync all awards / templates which have the default background image... However, this happens on demand, right? So... let's say that someone creates 10 templates with the default image. Then they change the default image and they *do not sync* (perhaps because they forget to or perhaps because they *do not want* to update the image on those 10 certificates.

Now lets say they create 10 more. We now have 10 templates with the very first original default image. Then 10 more with the updated user-defined new background image. 

Now they change it a 3rd time and they *do* sync... What does the user expect here? All 20 templates to update or only the most recent 10?

**Comment by manananomas** [comment]
IMHO - I believe a lot of our users are looking to update their look when they change the background... and by extension would accept the new look on an older cert. Where I see the difficulty is when a course creator gets creative and uses time markers in their artwork -- i.e. 2021 Certification... then if 2023 rolls around and the student needs to re-print an older Cert, it would be mislabeled by the background as an updated 2023. I hope that made sense. 

There is also the option to print the Certificates to pdf, so maybe leave it the way it is but also create documentation that encourages people to save pdf's of their existing certs?

**Comment by nrherron92** [comment]
I'm for: B) Don't store default images as the featured image and always pull the image from the option. 

The way I understand it is if you set a default, and then use the default in your templates (without changing it to an individual featured image) you expect the certificates to pull from the default. So if you change the default image it would change it all the places you're using the default.

**Comment by pondermatic** [comment]
Here's a good reason to NOT update backgrounds on awarded certificates when the default background image is changed in LifterLMS -> Engagements settings. Some organizations change branding/logos over time.  There are legal and marketing implications for overwriting the images that were appropriate at the time they were awarded.  For example, ownership of a company changing, trademarks, or brand standards.

This was suggested by my wife who was an education coordinator with LMS administration duties.

~In the short term, I prefer option A.~ In the long term, the user should be able to choose.

EDIT: If the background is expected to change because of the year or logos or something, it should be set as the template's featured image. The default background should be used on any awarded certificate that doesn't have an overriding featured image. I VOTE FOR B.

**Comment by chrisbadgett** [comment]
If a default background aka. "placeholder" is used, a default may be updated to a new placeholder in the future.

If a site admin would like users to create a non fungible version of their certificate built on top of the placeholder, they should encourage the user to "print" it or use LifterLMS PDFs to pull it out of the site as a PDF.

I like the idea of the alert explaining what happens when default background is changed:

"Used as the certificate background image for any certificate templates or awards which do not specify their own background image. Changing this background image will automatically update any existing certificate template or award which does not have a background image specified."

**Comment by pondermatic** [comment]
> a non fungible version of their certificate

Is adding blockchain technology to LifterLMS engagements a feature request? :)

**Comment by magicemilymiddleton** [comment]
I'll vote for B, because that's how themes work with customizer settings and language can fix it and it's not a big deal to me either way. (we have had a few people note that they'd like it to update upon the default)

**Comment by thomasplevy** [comment]
Alright we're going to go with B then. You've lost the popularity contest @eri-trabiccolo 

Additionally, I'm going to adjust the language from "Default" to "Placeholder" as per @chrisbadgett comment. I think calling it a placeholder as opposed to a "default" is a minor change but carries a lot of weight if you consider the difference between the two words.

**Comment by thomasplevy** [comment]
Also of note: 

> EDIT: If the background is expected to change because of the year or logos or something, it should be set as the template's featured image.

This doesn't help a user who didn't consider that theyd need to change their image but it's what we should be thinking about when we consider how this is supposed to work.

If you *don't want the default image to display* don't use the default image on your certificates.

**Comment by thomasplevy** [comment]
How's this? 

![Screenshot_2022-03-01_15-26-20](https://user-images.githubusercontent.com/1290739/156259329-3995f42f-2f8a-4aa2-beb8-d069eaca68b7.png)
:

**Comment by pondermatic** [comment]
Change "A default image used for certificates template or award which does not specify an image." to "A default image used for certificate templates or awards which do not specify an image.".

Or maybe just change "certificates" to "certificate".

**Comment by thomasplevy** [comment]
Good call. Closing!
